cluster/images/etcd-version-monitor/etcd-version-monitor.go

```diff
@@ -0,0 +1,275 @@
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package main
+
+import (
+	"encoding/json"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"net/http"
+	"os"
+	"os/exec"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/common/log"
+	"github.com/spf13/pflag"
+)
+
+// Initialize the prometheus instrumentation and client related flags.
+var (
+	listenAddress				string
+	metricsPath					string
+	etcdVersionScrapeURI		string
+	etcdMetricsScrapeURI		string
+	scrapeTimeout				time.Duration
+)
+
+func registerFlags(fs *pflag.FlagSet) {
+	fs.StringVar(&listenAddress, "listen-address", ":9101", "Address to listen on for serving prometheus formatted metrics")
+	fs.StringVar(&metricsPath, "metrics-path", "/metrics", "Path under which prometheus formatted metrics are to be served")
+	fs.StringVar(&etcdVersionScrapeURI, "etcd-version-scrape-uri", "http://localhost:2379/version", "URI to scrape etcd version info")
+	fs.StringVar(&etcdMetricsScrapeURI, "etcd-metrics-scrape-uri", "http://localhost:2379/metrics", "URI to scrape etcd metrics")
+	fs.DurationVar(&scrapeTimeout, "scrape-timeout", 15*time.Second, "Timeout for trying to get stats from etcd")
+}
+
+const (
+	namespace = "etcd"	// For prefixing prometheus metrics
+)
+
+// Initialize prometheus metrics to be exported.
+var (
+	etcdVersionFetchCount 	= prometheus.NewCounterVec(
+		prometheus.CounterOpts{
+			Namespace: namespace,
+			Name: "version_info_fetch_count",
+			Help: "Number of times etcd's version info was fetched, labeled by etcd's server binary and cluster version",
+		},
+		[]string{"server_version", "cluster_version"})
+	etcdVersionedDataSize 	= prometheus.NewGaugeVec(
+		prometheus.GaugeOpts{
+			Namespace: namespace,
+			Name: "versioned_data_size",
+			Help: "Size of etcd data in a given backend version (>=3), labeled by backend version, unit for size and etcdctl version used",
+		},
+		[]string{"backend_version", "unit", "etcdctl_version"})
+	etcdGRPCRequestsTotal	= prometheus.NewCounterVec(
+		prometheus.CounterOpts{
+			Namespace: namespace,
+			Name: "grpc_requests_total",
+			Help: "Counter of received grpc requests, labeled by grpc method and grpc service names",
+		},
+		[]string{"grpc_method", "grpc_service"})
+)
+
+// Struct for unmarshalling the json response from etcd's /version endpoint.
+type EtcdVersion struct {
+	ServerVersion		string `json:"etcdserver"`
+	ClusterVersion		string `json:"etcdcluster"`
+}
+
+// Function for periodically fetching etcd version info.
+func getVersion() {
+	// Create the get request for the etcd endpoint.
+	req, err := http.NewRequest("GET", etcdVersionScrapeURI, nil)
+	if err != nil {
+		log.Fatal("NewRequest: ", err)
+		return
+	}
+
+	for {
```

Comment:
It would be a better pattern for this function to just to a single request, and then do the loop over time in a separate function.